src/main/java/ai/elimu/rest/v2/crowdsource/ContributorsRestControllerWeb3.java

```diff
@@ -0,0 +1,174 @@
+            jsonObject.put("result", "error");
+            jsonObject.put("errorMessage", "Missing providerIdWeb3");
+            response.setStatus(HttpStatus.BAD_REQUEST.value());
+            
+            String jsonResponse = jsonObject.toString();
+            logger.info("jsonResponse: " + jsonResponse);
+            return jsonResponse;
+        }
+        
+        String providerIdWeb3Signature = contributorJSONObject.getString("providerIdWeb3Signature");
+        logger.info("providerIdWeb3Signature: " + providerIdWeb3Signature);
+        if (StringUtils.isBlank(providerIdWeb3Signature)) {
+            jsonObject.put("result", "error");
+            jsonObject.put("errorMessage", "Missing providerIdWeb3Signature");
+            response.setStatus(HttpStatus.BAD_REQUEST.value());
+            
+            String jsonResponse = jsonObject.toString();
+            logger.info("jsonResponse: " + jsonResponse);
+            return jsonResponse;
+        }
+        
+        // Check if the signature is valid
+        if (!isSignatureValid(providerIdWeb3, providerIdWeb3Signature, SIGNATURE_MESSAGE)) {
+            jsonObject.put("result", "error");
+            jsonObject.put("errorMessage", "Invalid signature");
+            response.setStatus(HttpStatus.BAD_REQUEST.value());
+            
+            String jsonResponse = jsonObject.toString();
+            logger.info("jsonResponse: " + jsonResponse);
+            return jsonResponse;
+        }
+        logger.info("Valid signature ✍️");
+        
+        // Check if a Contributor with this ETH address already exists in the database
+        Contributor existingContributor = contributorDao.readByProviderIdWeb3(providerIdWeb3);
+        logger.info("existingContributor: " + existingContributor);
+        if (existingContributor == null) {
+            // Store new Contributor in database
+            Contributor contributor = new Contributor();
+            contributor.setProviderIdWeb3(providerIdWeb3);
+            contributor.setEmail(providerIdWeb3 + "@ethmail.cc");
+            contributor.setRegistrationTime(Calendar.getInstance());
+            contributor.setRoles(new HashSet<>(Arrays.asList(Role.CONTRIBUTOR)));
+            contributorDao.create(contributor);
+            logger.info("Contributor " + contributor.getProviderIdWeb3() + " was created in the database");
+            
+            jsonObject.put("result", "success");
+            jsonObject.put("successMessage", "The Contributor was stored in the database");
+        } else {
+            jsonObject.put("result", "warning");
+            jsonObject.put("warningMessage", "The Contributor has already been stored in the database");
+        }
+        
+        String jsonResponse = jsonObject.toString();
+        logger.info("jsonResponse: " + jsonResponse);
+        return jsonResponse;
+    }
+    
+    private boolean isSignatureValid(final String address, final String signature, final String message) {
```

Comment:
As this function is used in [here](https://github.com/elimu-ai/webapp/blob/7e19a82e3ac46280fdb8fa9753ad35a0b66e61c6/src/main/java/ai/elimu/web/SignOnControllerWeb3.java#L125) too, it would be nice to moving this to a separate utils file to maintain a more integrated codebase.

Reply:
@hhio618 Yes, good suggestion! Fixed: https://github.com/elimu-ai/webapp/pull/1560/commits/a57f2bac7b60503ab04357a4b18d5e1c3125bbcb